### ISSUE TITLE
subgraph append

### ISSUE THREAD

**Issue by mef** [graphology/graphology-utils#6]
Consider the following use case:

A large dynamic graph is maintained, and only a subset of this graph is displayed to the user.

At a given moment, `subgraph()` is used to extract a small portion of a large dynamic graph (based on some metric). Some attributes are only computed on the subgraph, for performance reason (e.g. layout).

Later, we want to update the subgraph based on changes which have occurred on the main graph. 

Attributes need to be re-computed on the subgraph. For the nodes which remain present in the subgraph, it makes sense to re-use the values already determined.

If we overwrite the subgraph by calling `subgraph()` again, then the attributes will be lost. This is a drawback for attributes like positions in a force atlas 2 layout, because the nodes coordinates might shift totally, whereas it would be preferable to keep nodes close to their initial positions.

Hence, an `append` mode of subgraph creation is necessary, where attributes of the subgraph are kept (only those not present in the main graph).

Would you see such a feature fit inside `graphology-utils`, or should I implement it in a separate repository?

**Comment by Yomguithereal** [comment]
Hello @mef, this use case is the one of having a subgraph "view" on the graph, no? People usually rely on a class to do so then. This is an interesting issue and I am not sure to know the "best" way to proceed. Can you precise your use case or point me toward some kind of UI demonstrating it so I can assess a direction to take?

**Comment by mef** [comment]
Here's the use case:

Tweets are being extracted in real time via the streaming API, and a graph of hashtags used together in each tweet is updated continuously as data flows in.

The graph quickly grows to thousands of nodes.

In the front-end, a graph chart shows only the top 100 most popular nodes (based on degree), rather than the whole dataset. 

Because graph is updated in real time, it is too much overhead to compute FA2 layout to the whole graph over and over again. Only the subgraph of most popular nodes needs x / y coordinates.

Is the use case more clear now?

**Comment by Yomguithereal** [comment]
Ok, I see now. In this case, wouldn't it fit your use case more if the FA2 layout had an option to run only on a subset of your graph, as indicated by the list of top nodes?

**Comment by mef** [comment]
Yes, such a feature would indeed fit to the described use case.

I will need to do other processing on the subgraph (or view), such as computing Louvain modularity.

**Comment by Yomguithereal** [comment]
Hum... In that case I guess a subgraph view would be preferable indeed. This view would only need to implement the same interface as the whole graph so you can apply algorithm on it. But it poses an interesting indexing challenge and some design question about the interrelation between the graph and its subgraphs (regarding mutation, for instance). E.g. indexing only the set of the subgraph's nodes is quite cheap but can have a cost on edge methods if you need to filter irrelevant edges, while subindexing relevant edges can boost edge methods at the expense of memory. What do you think?

**Comment by mef** [comment]
I agree that having a concept of view on top the graph is more elegant than duplicating data in a separate sub-graph instance.
I am not sure about the impact on indexing and performance, not being very familiar with Graphology's internals yet.

The feature is a nice-to-have for me at the moment. I first considered doing the following as a workaround:

Compute FA2 layout on the subgraph, but assign the `x` and `y` attribute values to the nodes of the main graph.

After experimenting, the layout will be more meaningful when computed on the whole graph, rather than only on the subgraph. So I'll compute it for the whole graph, in near-real-time.
